File: 2020/2.1.rb
require_relative 'helpers.rb'

module DayTwoPartOne
  module_function

  def solve(file)
    inputs = prepare_inputs(file)
    find_valid_passwords(inputs)
  end

  def find_valid_passwords(inputs)
    inputs.reduce(0) do |total, value|
      min,
      max,
      letter,
      password = value.values_at(:min, :max, :letter, :password)
      letter_count = password.count(letter)
      total += 1 if letter_count >= min && letter_count <= max
      total
    end
  end

  def prepare_inputs(file)
    Helpers
      .split_inputs_by_line(Helpers.read_file(file))
      .map do |input|
        ## 2-9 c: ccccccccc
        inputs = input.split
        {
          min: inputs[0].split('-')[0].to_f,
          max: inputs[0].split('-')[1].to_f,
          letter: inputs[1].delete_suffix(':'),
          password: inputs[2]
        }
      end
  end
end

# puts DayTwoPartOne.solve(__dir__ + '/2test.txt')
puts DayTwoPartOne.solve(__dir__ + '/2.txt')

File: 2022/2.1.js
const fs = require('fs');

function solve(input) {
    return parseGames(input).reduce(addGame, 0)
}

function addGame(acc = 0, input) {
    let gameScore = 0;
    gameScore += input[1]
    if (input[0] == input[1]) {
        gameScore += 3
    } else {
        switch (input[0]) {
            case 1:
                gameScore += input[1] == 2 ? 6 : 0
                break;
            case 2:
                gameScore += input[1] == 3 ? 6 : 0
                break;
            default: 3
                gameScore += input[1] == 1 ? 6 : 0
        }
    }


    return acc + gameScore
}


function parseGames(input) {
    return input.split("\n").map(x => x.split(' ').map(parseInput))
}

function parseInput(input) {
    switch (input) {
        case "A":
        case "X":
            return 1
        case "B":
        case "Y":
            return 2
        default: // C, Z
            return 3
    }
}
console.log(solve(fs.readFileSync(__dirname + '/2test.txt', 'utf8')) == 15);
console.log(solve(fs.readFileSync(__dirname + '/2.txt', 'utf8')));
